File: RoombaPi.Client/src/app/bumper/services/bumper.service.ts
import { Injectable } from "@angular/core";

import { environment } from "src/environments/environment";

import { BumperServiceClient } from "@app/grpc/bumper";
import { Empty } from "google-protobuf/google/protobuf/empty_pb";

import { LightBumperStatus } from "../models/lightBumperStatus";
import { BumperStatus } from "../models/bumperStatus";

@Injectable()
export class BumperService {
  private readonly client: BumperServiceClient;

  constructor() {
    this.client = new BumperServiceClient(environment.grpcServer);
  }

  public getBumpsStatus(): Promise<BumperStatus> {
    return new Promise<BumperStatus>((resolve, reject) => {
      this.client.getBumpsStatus(new Empty(), null, (error, response) => {
        if (error) {
          return reject(error);
        }

        resolve(response.toObject());
      });
    });
  }

  public getLightBumpersStatus(): Promise<LightBumperStatus> {
    return new Promise<LightBumperStatus>((resolve, reject) => {
      this.client.getLightBumpersStatus(new Empty(), null, (error, response) => {
        if (error) {
          return reject(error);
        }

        resolve(response.toObject());
      });
    });
  }

  public isVirtualWallPresent(): Promise<boolean> {
    return new Promise<boolean>((resolve, reject) => {
      this.client.isVirtualWallPresent(new Empty(), null, (error, response) => {
        if (error) {
          return reject(error);
        }

        resolve(response.getVirtualWallPresent());
      });
    });
  }
}

File: RoombAPI.Server/Services/BumperService.cs
namespace RoombaPi.Server.Services
{
  using System.Threading.Tasks;
  using Create2.Sensors.Environment;
  using global::Grpc.Core;
  using Google.Protobuf.WellKnownTypes;
  using RoombaPi.Grpc;

  public class BumperService : Bumper.BumperBase
  {
    private readonly IEnvironmentSensor environmentSensor;

    public BumperService(IEnvironmentSensor environmentSensor)
    {
      this.environmentSensor = environmentSensor;
    }

    public override async Task<BumpsStatusResponse> GetBumpsStatus(Empty request, ServerCallContext context)
    {
      var bumpsAndWheelDrops = await this.environmentSensor.GetBumpsAndWheelDrops();
      return new BumpsStatusResponse
      {
        Left = (bumpsAndWheelDrops & BumpsAndWheelDrops.BumpLeft) == BumpsAndWheelDrops.BumpLeft,
        Right = (bumpsAndWheelDrops & BumpsAndWheelDrops.BumpRight) == BumpsAndWheelDrops.BumpRight
      };
    }

    public override async Task<LightBumpersStatusResponse> GetLightBumpersStatus(Empty request,
      ServerCallContext context)
    {
      var lightBumperDetections = await this.environmentSensor.GetLightBumperDetections();
      var response = new LightBumpersStatusResponse();
      response.Left = new LightBumperInfo
      {
        Active = (lightBumperDetections & LightBumperDetections.Left) == LightBumperDetections.Left,
        Signal = await this.environmentSensor.GetLightBumpLeftSignal()
      };

      response.FrontLeft = new LightBumperInfo
      {
        Active = (lightBumperDetections & LightBumperDetections.FrontLeft) == LightBumperDetections.FrontLeft,
        Signal = await this.environmentSensor.GetLightBumpFrontLeftSignal()
      };

      response.CenterLeft = new LightBumperInfo
      {
        Active = (lightBumperDetections & LightBumperDetections.CenterLeft) == LightBumperDetections.CenterLeft,
        Signal = await this.environmentSensor.GetLightBumpCenterLeftSignal()
      };

      response.Right = new LightBumperInfo
      {
        Active = (lightBumperDetections & LightBumperDetections.Right) == LightBumperDetections.Right,
        Signal = await this.environmentSensor.GetLightBumpLeftSignal()
      };

      response.FrontRight = new LightBumperInfo
      {
        Active = (lightBumperDetections & LightBumperDetections.FrontRight) == LightBumperDetections.FrontRight,
        Signal = await this.environmentSensor.GetLightBumpFrontLeftSignal()
      };

      response.CenterRight = new LightBumperInfo
      {
        Active = (lightBumperDetections & LightBumperDetections.CenterRight) == LightBumperDetections.CenterRight,
        Signal = await this.environmentSensor.GetLightBumpCenterLeftSignal()
      };

      return response;
    }

    public override async Task<IsVirtualWallPresentResponse> IsVirtualWallPresent(Empty request,
      ServerCallContext context)
    {
      return new IsVirtualWallPresentResponse
      {
        VirtualWallPresent = await this.environmentSensor.GetVirtualWall()
      };
    }
  }
}